### PR TITLE
engine: add H2 ping config API

### PR DESCRIPTION
**Pull request by junr03**
Description: adds engine builder and config plumbing to setup h2 ping support
Risk Level: low - exposed API. Default config is no-op in Envoy
Testing: added tests
Docs Changes: pending if reviewers agree with API

### CODE REVIEW

File: library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java [hunk 1]
```diff
@@ -118,6 +131,12 @@ String resolveTemplate(final String templateYAML, final String platformFilterTem
         .append(String.format("- &dns_fail_max_interval %ss\n", dnsFailureRefreshSecondsMax))
         .append(String.format("- &dns_query_timeout %ss\n", dnsQueryTimeoutSeconds))
         .append(String.format("- &dns_preresolve_hostnames %s\n", dnsPreresolveHostnames))
+        .append(String.format("- &h2_connection_keepalive_interval %ss\n",
+                              h2ConnectionKeepaliveIntervalSeconds))
+        .append(String.format("- &h2_connection_keepalive_idle_interval %sms\n",
```
Comment (junr03) [inline library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java:136]: I dont think `ms` works automagically. Will update tomorrow.

Comment (goaway) [reply library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java:136]: Only `s` works with the proto parser.

File: library/cc/engine_builder.cc [hunk 2]
```diff
@@ -54,6 +54,25 @@ EngineBuilder::addDnsPreresolveHostnames(const std::string& dns_preresolve_hostn
...
+  this->h2_connection_keepalive_idle_interval_milliseconds_ =
+      h2_connection_keepalive_idle_interval_milliseconds;
+  return *this;
+}
+
+EngineBuilder&
+EngineBuilder::addH2ConnectionKeepaliveTimeoutSeconds(int h2_connection_keepalive_timeout_seconds) {
+  this->h2_connection_keepalive_timeout_seconds_ = h2_connection_keepalive_timeout_seconds;
+  return *this;
+}
```
Comment (goaway) [inline library/cc/engine_builder.cc:74]: I'm actually not certain we want to expose these values at this layer. If this ends up being our primary mechanism to detect zombie connections, overriding this configuration seems like a pretty major footgun for a naive user of the library.

File: library/common/config/config.cc [hunk 3]
```diff
@@ -39,6 +39,9 @@ const std::string config_header = R"(
 - &dns_fail_max_interval 10s
 - &dns_query_timeout 25s
 - &dns_preresolve_hostnames []
+- &h2_connection_keepalive_interval 0s
+- &h2_connection_keepalive_idle_interval 0s
+- &h2_connection_keepalive_timeout 5s
```
Comment (goaway) [inline library/common/config/config.cc:44]: It looks like these still need to be referenced in the cluster config.

Comment (junr03) [reply library/common/config/config.cc:44]: Yep, I didn't want to finish wiring until we settled on interface.

File: library/cc/engine_builder.cc (hunk 2 above)
Comment (junr03) [reply library/cc/engine_builder.cc:74]: @goaway I wasn't aware that was your preference. I described that I was exposing all the way at the Engine API. What would your preference be? If we dont expose them at the engine builder layer then we wouldn't be able to experiment with settings internally at Lyft.

Comment (junr03) [reply library/cc/engine_builder.cc:68]: going to set up default API in a subsequent PR. per offline discussion.

File: library/common/config/config.cc [hunk 4]
```diff
@@ -39,6 +39,8 @@ const std::string config_header = R"(
 - &dns_fail_max_interval 10s
 - &dns_query_timeout 25s
 - &dns_preresolve_hostnames []
+- &h2_connection_keepalive_idle_interval 100000s
```
Comment (goaway) [inline library/common/config/config.cc:42]: Probably should be 0 here just so it's disabled in vanilla config.

File: library/cc/engine_builder.h [hunk 5]
```diff
@@ -53,6 +57,8 @@ class EngineBuilder {
   int dns_failure_refresh_seconds_max_ = 10;
   int dns_query_timeout_seconds_ = 25;
   std::string dns_preresolve_hostnames_ = "[]";
+  int h2_connection_keepalive_idle_interval_milliseconds_ = 0;
+  int h2_connection_keepalive_timeout_seconds_ = 5;
```
Comment (goaway) [inline library/cc/engine_builder.h:61]: I'd suggest 10 as a starting point.

File: library/common/config/config.cc (hunk 4 above)
Comment (junr03) [reply library/common/config/config.cc:42]: unfortunately the proto constraint to at least 1ms.

File: library/common/config/config.cc [hunk 6]
```diff
@@ -39,6 +39,8 @@ const std::string config_header = R"(
...
+- &h2_connection_keepalive_timeout 5s
```
Comment (goaway) [inline library/common/config/config.cc:43]: 10

File: library/cc/engine_builder.h (hunk 5 above)
Comment (junr03) [reply library/cc/engine_builder.h:61]: oh right, will update.

File: library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt [hunk 7]
```diff
@@ -35,6 +35,8 @@ open class EngineBuilder(
   private var dnsFailureRefreshSecondsMax = 10
   private var dnsQueryTimeoutSeconds = 25
   private var dnsPreresolveHostnames = "[]"
+  private var h2ConnectionKeepaliveIdleIntervalMilliseconds = 0
+  private var h2ConnectionKeepaliveTimeoutSeconds = 5
```
Comment (goaway) [inline library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:39]: 10

File: library/swift/EngineBuilder.swift [hunk 8]
```diff
@@ -21,6 +21,8 @@ open class EngineBuilder: NSObject {
   private var dnsFailureRefreshSecondsMax: UInt32 = 10
   private var dnsQueryTimeoutSeconds: UInt32 = 25
   private var dnsPreresolveHostnames: String = "[]"
+  private var h2ConnectionKeepaliveIdleIntervalMilliseconds: UInt32 = 0
+  private var h2ConnectionKeepaliveTimeoutSeconds: UInt32 = 5
```
Comment (goaway) [inline library/swift/EngineBuilder.swift:25]: 10

**Comment by goaway** [comment]
Other than some nits about the defaults, I think this looks good. An integration test would be amazing, but that doesn't *need* to be part of this PR.

**Comment by goaway** [comment]
Also, suggest merging main, to make sure you've picked up the new RBE configs for CI.

**Review by goaway — APPROVED** [review]

**Comment by goaway** [comment]
Thanks @junr03!

**Pull request merged**